jetty-websocket/websocket-javax-common/src/main/java/org/eclipse/jetty/websocket/javax/common/messages/AbstractDecodedMessageSink.java

```diff
@@ -0,0 +1,99 @@
+//
+// ========================================================================
+// Copyright (c) 1995-2020 Mort Bay Consulting Pty Ltd and others.
+//
+// This program and the accompanying materials are made available under
+// the terms of the Eclipse Public License 2.0 which is available at
+// https://www.eclipse.org/legal/epl-2.0
+//
+// This Source Code may also be made available under the following
+// Secondary Licenses when the conditions for such availability set
+// forth in the Eclipse Public License, v. 2.0 are satisfied:
+// the Apache License v2.0 which is available at
+// https://www.apache.org/licenses/LICENSE-2.0
+//
+// SPDX-License-Identifier: EPL-2.0 OR Apache-2.0
+// ========================================================================
+//
+
+package org.eclipse.jetty.websocket.javax.common.messages;
+
+import java.lang.invoke.MethodHandle;
+import java.util.List;
+import java.util.stream.Collectors;
+import javax.websocket.Decoder;
+
+import org.eclipse.jetty.util.Callback;
+import org.eclipse.jetty.websocket.core.CoreSession;
+import org.eclipse.jetty.websocket.core.Frame;
+import org.eclipse.jetty.websocket.javax.common.decoders.RegisteredDecoder;
+import org.eclipse.jetty.websocket.util.messages.MessageSink;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public abstract class AbstractDecodedMessageSink implements MessageSink
+{
+    protected final Logger _logger;
+    protected final CoreSession _coreSession;
+    protected final MethodHandle _methodHandle;
+    protected final MessageSink _messageSink;
```

Comment:
`protected` fields should be avoided.
Make them private and if really necessary, use package private methods to access them.

`_logger` should be duplicated in all subclasses so we know whether we are logging from binary or text subclasses.

`_methodHandle` is only used by subclasses, so better to have it private, with a method called by subclasses.

`_messageSink` is weird because it is called from the constructor (which typically must no be done - leaks `this` as a partially constructed object), and only used in `accept()`.
I would prefer subclasses implement `accept()` and each have its own messageSink - thoughts?

Reply:
To me the current structure makes the most sense, where to create a `DecodedMessageSink` all you need to do is implement the `getMessageSink` method.

Adding `accept()` to every version of the `DecodedMessageSink` is just going to duplicate the same method 4 times. What do you think about creating the `MessageSink` in the constructor and giving it to super. OR just passing in a Provider<MessageSink> to the super constructor?